### ISSUE TITLE
是否稀疏化训练

### ISSUE THREAD

**Issue by muzili-pengyang** [zhahoi/yolov11_prune_distillation#3]
if self.prune and self.pruner is None:
            example_inputs = torch.randn(1, 3, self.args.imgsz, self.args.imgsz)
            ignored_layers = []
            for m in self.model.modules():
                if isinstance(m, (Detect, Attention, Pose)):
                    ignored_layers.append(m)
            self.pruner = tp.pruner.MagnitudePruner(
                self.model,
                example_inputs,
                importance = tp.importance.MagnitudeImportance(p=2),  # L2 norm pruning,
                iterative_steps=self.prune_iterative_steps,
                pruning_ratio=self.prune_ratio,
                ignored_layers=ignored_layers,
            )

        if self.prune and self.pruner is not None:
            example_inputs = torch.randn(1, 3, self.args.imgsz, self.args.imgsz)
            self.pruner.step()
            self.model = self.model.to(self.device)
            pruned_macs, pruned_nparams = tp.utils.count_ops_and_params(self.model, example_inputs.to(self.device))
            print(f"After pruning iter : MACs={pruned_macs / 1e9} G, #Params={pruned_nparams / 1e6} M, ")

            if str(path).endswith('.pt') and self.prune_load:
                weights, _ = attempt_load_one_weight(path)
                self.model.load(weights)
        else:
            self.model = self.model.to(self.device)

       在第一阶段稀疏化训练时，model.train(data=data, epochs=train_epochs, imgsz=imgsz, batch=batch, device=device, name=f"{name}_phase1", prune=False,sparse_training=True)，由于prune=False，不会构建pruner，之后进行训练
   就是启动sparse_training ，但pruner is None，也不会进入习稀疏化训练啊
          
           if self.sparse_training and self.pruner is not None:
                self.pruner.update_regularizer() # Init every epoch
            for i, batch in pbar:
                self.run_callbacks("on_train_batch_start")
                # Warmup
                ni = i + nb * epoch
                if ni <= nw:
                    xi = [0, nw]  # x interp
                    self.accumulate = max(1, int(np.interp(ni, xi, [1, self.args.nbs / self.batch_size]).round()))
                    for j, x in enumerate(self.optimizer.param_groups):
                        # Bias lr falls from 0.1 to lr0, all other lrs rise from 0.0 to lr0
                        x["lr"] = np.interp(
                            ni, xi, [self.args.warmup_bias_lr if j == 0 else 0.0, x["initial_lr"] * self.lf(epoch)]
                        )
                        if "momentum" in x:
                            x["momentum"] = np.interp(ni, xi, [self.args.warmup_momentum, self.args.momentum])

                # Forward
                with autocast(self.amp):
                    batch = self.preprocess_batch(batch)
                    loss, self.loss_items = self.model(batch)
                    self.loss = loss.sum()
                    if RANK != -1:
                        self.loss *= world_size
                    self.tloss = (
                        (self.tloss * i + self.loss_items) / (i + 1) if self.tloss is not None else self.loss_items
                    )
                    
                    # Added
                    if self.exist_dis:
                        distill_weight = (((1 - math.cos(i * math.pi / len(self.train_loader))) / 2) * (
                                    0.1 - 1) + 1) * 0.01
                        with torch.no_grad():
                            pred = self.Distillation(batch['img'])

                        self.d_loss = distillation_loss.get_loss()
                        self.d_loss *= distill_weight
                        if i == 0:
                            print("distillation_loss : ", self.d_loss, '-----------------', "loss : ", self.loss,
                                  '-----------------')
                        self.loss += self.d_loss

                # Backward
                self.scaler.scale(self.loss).backward()

                if self.sparse_training and self.pruner is not None:
                    self.pruner.regularize(self.model, self.loss) # After loss.backward(), Before optimizer.step()
